File: Client_Root/Client/Assets/Scripts/Network/Messages/ToRoom/GameResultToR.cs
using FlatBuffers;
using System.Collections.Generic;

public class GameResultToR : IMessage
{
    public const ushort MESSAGE_ID = MessageID.GameResultToR_ID;

    public List<PlayerRankInfo> m_listPlayerRankInfo = new List<PlayerRankInfo>();

    public override ushort GetID()
    {
        return MESSAGE_ID;
    }

    public override IMessage Clone()
    {
        return null; 
    }

    public override byte[] Serialize()
    {
        FBS.GameResultToR.StartPlayerRanksVector(m_Builder, m_listPlayerRankInfo.Count);
        foreach(PlayerRankInfo playerRankInfo in m_listPlayerRankInfo)
        {
            FBS.Data.PlayerRankInfo.CreatePlayerRankInfo(m_Builder, playerRankInfo.m_nPlayerIndex, playerRankInfo.m_nRank, playerRankInfo.m_fHeight);
        }
        var playerRanks = m_Builder.EndVector();

        FBS.GameResultToR.StartGameResultToR(m_Builder);
        FBS.GameResultToR.AddPlayerRanks(m_Builder, playerRanks);
        var data = FBS.GameResultToR.EndGameResultToR(m_Builder);

        m_Builder.Finish(data.Value);

        return m_Builder.SizedByteArray();
    }

    public override bool Deserialize(byte[] bytes)
    {
        var buf = new ByteBuffer(bytes);

        var data = FBS.GameResultToR.GetRootAsGameResultToR(buf);

        for (int i = 0; i < data.PlayerRanksLength; ++i)
        {
            m_listPlayerRankInfo.Add(new PlayerRankInfo(data.GetPlayerRanks(i)));
        }

        return true;
    }

    public override void OnReturned()
    {
        m_listPlayerRankInfo.Clear();
    }
}

File: Server_Root/Server/RoomServer/Messages/ToRoom/GameResultToR.cpp
#include "stdafx.h"
#include "GameResultToR.h"

GameResultToR::GameResultToR()
{
}

GameResultToR::~GameResultToR()
{
}

unsigned short GameResultToR::GetID()
{
	return MESSAGE_ID;
}

IMessage* GameResultToR::Clone()
{
	GameResultToR* pClone = new GameResultToR();

	for (int i = 0; i < m_vecPlayerRankInfo.size(); ++i)
	{
		pClone->m_vecPlayerRankInfo.push_back(FBS::Data::PlayerRankInfo(m_vecPlayerRankInfo[i].PlayerIndex(), m_vecPlayerRankInfo[i].Rank(), m_vecPlayerRankInfo[i].Height()));
	}

	return pClone;
}

const char* GameResultToR::Serialize(int* pLength)
{
	auto playerRankInfos = m_Builder.CreateVectorOfStructs(m_vecPlayerRankInfo);

	FBS::GameResultToRBuilder data_builder(m_Builder);
	data_builder.add_PlayerRanks(playerRankInfos);
	auto data = data_builder.Finish();

	m_Builder.Finish(data);

	*pLength = m_Builder.GetSize();

	return (char*)m_Builder.GetBufferPointer();
}

bool GameResultToR::Deserialize(const char* pChar)
{
	auto data = flatbuffers::GetRoot<FBS::GameResultToR>((const void*)pChar);

	for (int i = 0; i < data->PlayerRanks()->size(); ++i)
	{
		auto playerRank = data->PlayerRanks()->Get(i);
		m_vecPlayerRankInfo.push_back(FBS::Data::PlayerRankInfo(playerRank->PlayerIndex(), playerRank->Rank(), playerRank->Height()));
	}

	return true;
}